File: Src/Compilers/VisualBasic/Source/Errors/Location.vb
Imports Roslyn.Compilers.Common
Imports Roslyn.Compilers.Internal

Namespace Roslyn.Compilers.VisualBasic
    ''' <summary>
    ''' Represents a location in source code or metadata. Used for the location of diagnostics and symbols.
    ''' </summary>
    Public MustInherit Class Location
        Implements ILocation

        Public MustOverride ReadOnly Property Kind As LocationKind Implements ILocation.Kind

        ''' <summary>
        ''' Indicates if the location refers to a particular place in source code.
        ''' </summary>
        Public MustOverride ReadOnly Property InSource As Boolean Implements ILocation.InSource

        ''' <summary>
        ''' Indicates if the location refers to metadata.
        ''' </summary>
        Public MustOverride ReadOnly Property InMetadata As Boolean Implements ILocation.InMetadata

        ''' <summary>
        ''' The location within the syntax tree that this error is associated with.
        ''' Throws an InvalidOperationException unless InSource is true.
        ''' </summary>
        Public MustOverride ReadOnly Property SourceSpan As TextSpan Implements ILocation.SourceSpan

        ''' <summary>
        ''' The syntax tree this error is located in. Throws
        ''' an InvalidOperationException unless InSource is true.
        ''' </summary>
        Public MustOverride ReadOnly Property SourceTree As SyntaxTree

        Private ReadOnly Property SourceTree1 As CommonSyntaxTree Implements ILocation.SourceTree
            Get
                Return Me.SourceTree
            End Get
        End Property

        ''' <summary>
        ''' Returns the metadata module the error is associated with. 
        ''' Throws an InvalidOperationException unless InMetadata is true.
        ''' </summary>
        Public MustOverride ReadOnly Property MetadataModule As ModuleSymbol

        ''' <summary>
        ''' Gets the location in terms of file name/line/column.
        ''' Throws an InvalidOperationException unless InSource is true.
        ''' </summary>
        ''' <param name="usePreprocessorDirectives">If true, the filename/line/column
        ''' reported takes into account #line directivees. If false, #line directives
        ''' are ignored.</param>
        ''' <returns></returns>
        Public MustOverride Function GetLineSpan(usePreprocessorDirectives As Boolean) As FileLinePositionSpan Implements ILocation.GetLineSpan

        ' Derived classes should provide value equality semantics.
        Public MustOverride Overrides Function [Equals](obj As Object) As Boolean
        Public MustOverride Overrides Function GetHashCode() As Integer
    End Class

    ''' <summary>
    ''' A program location in source code.
    ''' </summary>
    Friend Class SourceLocation
        Inherits Location

        Private ReadOnly _syntaxTree As SyntaxTree
        Private ReadOnly _span As TextSpan

        Public Overrides ReadOnly Property Kind As LocationKind
            Get
                Return LocationKind.SourceFile
            End Get
        End Property


        Public Sub New(syntaxTree As SyntaxTree,
                       span As TextSpan)
            Contract.ThrowIfNull(syntaxTree)

            _syntaxTree = syntaxTree
            _span = span
        End Sub

        Public Sub New(syntaxTree As SyntaxTree,
                       node As SyntaxNode)
            Me.New(syntaxTree, node.Span)
        End Sub

        Public Sub New(syntaxTree As SyntaxTree,
                       token As SyntaxToken)
            Me.New(syntaxTree, token.Span)
        End Sub

        Public Sub New(syntaxTree As SyntaxTree,
                       nodeOrToken As SyntaxNodeOrToken)
            Me.New(syntaxTree, nodeOrToken.Span)
        End Sub

        Public Sub New(syntaxTree As SyntaxTree,
                       trivia As SyntaxTrivia)
            Me.New(syntaxTree, trivia.Span)
        End Sub

        Public Sub New(syntaxRef As SyntaxReference)
            Me.New(syntaxRef.SyntaxTree, syntaxRef.Span)

            ' If we're using a syntaxref, we don't have a node in hand, so we couldn't get equality
            ' on syntax node, so associatedNode shouldn't be set. We never use this constructor
            ' when binding executable code anywhere, so it has no use.
        End Sub

        Public Overrides ReadOnly Property InMetadata As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property InSource As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property MetadataModule As ModuleSymbol
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides ReadOnly Property SourceSpan As TextSpan
            Get
                Return _span
            End Get
        End Property

        Public Overrides ReadOnly Property SourceTree As SyntaxTree
            Get
                Return _syntaxTree
            End Get
        End Property

        Public Overrides Function GetLineSpan(usePreprocessorDirectives As Boolean) As FileLinePositionSpan
            Return _syntaxTree.GetLineSpan(_span, usePreprocessorDirectives)
        End Function

        Public Overloads Function Equals(other As SourceLocation) As Boolean
            Return other IsNot Nothing AndAlso other._syntaxTree Is _syntaxTree AndAlso other._span.Equals(_span)
        End Function

        Public Overloads Overrides Function Equals(obj As Object) As Boolean
            Return Me.Equals(TryCast(obj, SourceLocation))
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Hash.Combine(_syntaxTree.GetHashCode(), _span.GetHashCode())
        End Function
    End Class

    ''' <summary>
    ''' A program location in speculative bound code. Right now, speculatively bound code
    ''' doesn't have a syntax tree associated with it. This probably will change in the future, but
    ''' for now this works.
    ''' </summary>
    Friend Class SpeculativeLocation
        Inherits Location

        Private ReadOnly _span As TextSpan

        Public Overrides ReadOnly Property Kind As LocationKind
            Get
                Return LocationKind.None  ' TODO: should be better kind than this?
            End Get
        End Property

        Public Sub New(span As TextSpan)
            _span = span
        End Sub

        Public Sub New(node As SyntaxNode)
            Me.New(node.Span)
            '_associatedNode = New WeakReference(Of SyntaxNode)(node)
            '_associateInParent = associateInParent
        End Sub

        Public Sub New(token As SyntaxToken)
            Me.New(token.Span)
            '_associatedNode = New WeakReference(Of SyntaxNode)(token.Parent)
            '_associateInParent = False
        End Sub

        Public Sub New(nodeOrToken As SyntaxNodeOrToken)
            Me.New(nodeOrToken.Span)
            'If nodeOrToken.IsNode Then
            '    _associatedNode = New WeakReference(Of SyntaxNode)(nodeOrToken.AsNode())
            'Else
            '    _associatedNode = New WeakReference(Of SyntaxNode)(nodeOrToken.AsToken().Parent)
            'End If

            '_associateInParent = False
        End Sub

        Public Sub New(trivia As SyntaxTrivia)
            Me.New(trivia.Span)
            '_associatedNode = New WeakReference(Of SyntaxNode)(trivia.Token.Parent)
            '_associateInParent = False
        End Sub

        Public Overrides ReadOnly Property InMetadata As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property InSource As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property MetadataModule As ModuleSymbol
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides ReadOnly Property SourceSpan As TextSpan
            Get
                Return _span
            End Get
        End Property

        Public Overrides ReadOnly Property SourceTree As SyntaxTree
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides Function GetLineSpan(usePreprocessorDirectives As Boolean) As FileLinePositionSpan
            Throw New InvalidOperationException()
        End Function

        Public Overloads Function Equals(other As SpeculativeLocation) As Boolean
            Return other IsNot Nothing AndAlso other._span.Equals(_span)
        End Function

        Public Overloads Overrides Function Equals(obj As Object) As Boolean
            Return Me.Equals(TryCast(obj, SpeculativeLocation))
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return _span.GetHashCode()
        End Function
    End Class

    ''' <summary>
    ''' A program location in metadata.
    ''' </summary>
    Friend NotInheritable Class MetadataLocation
        Inherits Location

        Private _module As ModuleSymbol

        Public Sub New([module] As ModuleSymbol)
            Contract.ThrowIfNull([module])
            _module = [module]
        End Sub

        Public Overrides ReadOnly Property Kind As LocationKind
            Get
                Return LocationKind.MetadataFile
            End Get
        End Property

        Public Overrides ReadOnly Property InSource As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property InMetadata As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property MetadataModule As ModuleSymbol
            Get
                Return _module
            End Get
        End Property

        Public Overrides ReadOnly Property SourceSpan As TextSpan
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides ReadOnly Property SourceTree As SyntaxTree
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides Function GetLineSpan(usePreprocessorDirectives As Boolean) As FileLinePositionSpan
            Throw New InvalidOperationException()
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return _module.GetHashCode()
        End Function

        Public Overrides Function Equals(obj As Object) As Boolean
            Dim other As MetadataLocation = TryCast(obj, MetadataLocation)
            Return (other IsNot Nothing) AndAlso (other._module Is Me._module)
        End Function
    End Class

    ''' <summary>
    ''' A class that represents no location at all. Useful for errors in command line options, for example.
    ''' </summary>
    ''' <remarks></remarks>
    Friend Class NoLocation
        Inherits Location

        Public Shared Singleton As New NoLocation

        Private Sub New()
            ' nothing to do.
        End Sub

        Public Overrides ReadOnly Property Kind As LocationKind
            Get
                Return LocationKind.None
            End Get
        End Property

        Public Overrides ReadOnly Property InSource As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property InMetadata As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property MetadataModule As ModuleSymbol
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides ReadOnly Property SourceSpan As TextSpan
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides ReadOnly Property SourceTree As SyntaxTree
            Get
                Throw New InvalidOperationException()
            End Get
        End Property

        Public Overrides Function GetLineSpan(usePreprocessorDirectives As Boolean) As FileLinePositionSpan
            Throw New InvalidOperationException()
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return &H58914756  ' arbitrary number, since all NoLocation's are equal
        End Function

        Public Overrides Function Equals(obj As Object) As Boolean
            Return TypeOf obj Is NoLocation
        End Function

    End Class

End Namespace

File: src/Compilers/Core/Portable/Diagnostic/Location.cs
﻿// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#nullable enable

using System;
using System.Diagnostics;
using Microsoft.CodeAnalysis.Symbols;
using Microsoft.CodeAnalysis.Text;

namespace Microsoft.CodeAnalysis
{
    /// <summary>
    /// A program location in source code.
    /// </summary>
    [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
    public abstract class Location
    {
        internal Location()
        {
        }

        /// <summary>
        /// Location kind (None/SourceFile/MetadataFile).
        /// </summary>
        public abstract LocationKind Kind { get; }

        /// <summary>
        /// Returns true if the location represents a specific location in a source code file.
        /// </summary>
        public bool IsInSource { get { return SourceTree != null; } }

        /// <summary>
        /// Returns true if the location is in metadata.
        /// </summary>
        public bool IsInMetadata { get { return MetadataModuleInternal != null; } }

        /// <summary>
        /// The syntax tree this location is located in or <c>null</c> if not in a syntax tree.
        /// </summary>
        public virtual SyntaxTree? SourceTree { get { return null; } }

        /// <summary>
        /// Returns the metadata module the location is associated with or <c>null</c> if the module is not available.
        /// </summary>
        /// <remarks>
        /// Might return null even if <see cref="IsInMetadata"/> returns true. The module symbol might not be available anymore, 
        /// for example, if the location is serialized and deserialized.
        /// </remarks>
        public IModuleSymbol? MetadataModule { get { return (IModuleSymbol?)MetadataModuleInternal?.GetISymbol(); } }

        internal virtual IModuleSymbolInternal? MetadataModuleInternal { get { return null; } }

        /// <summary>
        /// The location within the syntax tree that this location is associated with.
        /// </summary>
        /// <remarks>
        /// If <see cref="IsInSource"/> returns False this method returns an empty <see cref="TextSpan"/> which starts at position 0.
        /// </remarks>
        public virtual TextSpan SourceSpan { get { return default(TextSpan); } }

        /// <summary>
        /// Gets the location in terms of path, line and column.
        /// </summary>
        /// <returns>
        /// <see cref="FileLinePositionSpan"/> that contains path, line and column information.
        /// 
        /// Returns an invalid span (see <see cref="FileLinePositionSpan.IsValid"/>) if the information is not available.
        /// 
        /// The values are not affected by line mapping directives (#line in C# or #ExternalSource in VB).
        /// </returns>
        public virtual FileLinePositionSpan GetLineSpan()
        {
            return default(FileLinePositionSpan);
        }

        /// <summary>
        /// Gets the location in terms of path, line and column after applying source line mapping directives
        /// (<c>#line</c> in C# or <c>#ExternalSource</c> in VB).
        /// </summary>
        /// <returns>
        /// <see cref="FileLinePositionSpan"/> that contains file, line and column information,
        /// or an invalid span (see <see cref="FileLinePositionSpan.IsValid"/>) if not available.
        /// </returns>
        public virtual FileLinePositionSpan GetMappedLineSpan()
        {
            return default(FileLinePositionSpan);
        }

        // Derived classes should provide value equality semantics.
        public abstract override bool Equals(object? obj);
        public abstract override int GetHashCode();

        public override string ToString()
        {
            string result = Kind.ToString();
            if (IsInSource)
            {
                result += "(" + this.SourceTree?.FilePath + this.SourceSpan + ")";
            }
            else if (IsInMetadata)
            {
                if (this.MetadataModuleInternal != null)
                {
                    result += "(" + this.MetadataModuleInternal.Name + ")";
                }
            }
            else
            {
                var pos = GetLineSpan();
                if (pos.Path != null)
                {
                    // user-visible line and column counts are 1-based, but internally are 0-based.
                    result += "(" + pos.Path + "@" + (pos.StartLinePosition.Line + 1) + ":" + (pos.StartLinePosition.Character + 1) + ")";
                }
            }

            return result;
        }

        public static bool operator ==(Location? left, Location? right)
        {
            if (object.ReferenceEquals(left, null))
            {
                return object.ReferenceEquals(right, null);
            }

            return left.Equals(right);
        }

        public static bool operator !=(Location? left, Location? right)
        {
            return !(left == right);
        }

        protected virtual string GetDebuggerDisplay()
        {
            string result = this.GetType().Name;
            var pos = GetLineSpan();
            if (pos.Path != null)
            {
                // user-visible line and column counts are 1-based, but internally are 0-based.
                result += "(" + pos.Path + "@" + (pos.StartLinePosition.Line + 1) + ":" + (pos.StartLinePosition.Character + 1) + ")";
            }

            return result;
        }

        /// <summary>
        /// A location of kind LocationKind.None. 
        /// </summary>
        public static Location None { get { return NoLocation.Singleton; } }

        /// <summary>
        /// Creates an instance of a <see cref="Location"/> for a span in a <see cref="SyntaxTree"/>.
        /// </summary>
        public static Location Create(SyntaxTree syntaxTree, TextSpan textSpan)
        {
            if (syntaxTree == null)
            {
                throw new ArgumentNullException(nameof(syntaxTree));
            }

            return new SourceLocation(syntaxTree, textSpan);
        }

        /// <summary>
        /// Creates an instance of a <see cref="Location"/> for a span in a file.
        /// </summary>
        public static Location Create(string filePath, TextSpan textSpan, LinePositionSpan lineSpan)
        {
            if (filePath == null)
            {
                throw new ArgumentNullException(nameof(filePath));
            }

            return new ExternalFileLocation(filePath, textSpan, lineSpan);
        }
    }
}
